# Answering Business Questions using SQL

In this project, we will use the Chinook database.
The Chinook database is provided as a SQLite database file called chinook.db.  A copy of the database schema is below

![chinook schema](https://s3.amazonaws.com/dq-content/280/chinook-schema.svg)

In [4]:
import sqlite3
import pandas as pd

In [5]:
def run_query(my_query):
    with sqlite3.connect('chinook.db') as conn:
        query_result = pd.read_sql(my_query, conn)
    return query_result

def run_command(my_query):
    with sqlite3.connect('chinook.db') as conn:
        conn.isolation_level = None
        conn.execute(c)
        
def show_tables():
    kueri = "SELECT name, type FROM sqlite_master WHERE type IN ('table','view');"
    return run_query(kueri)

In [6]:
show_tables()

name   type
0            album  table
1           artist  table
2         customer  table
3         employee  table
4            genre  table
5          invoice  table
6     invoice_line  table
7       media_type  table
8         playlist  table
9   playlist_track  table
10           track  table
11     top_5_names   view

# Case 1 : Recommendation for new artist's albums

The Chinook record store has just signed a deal with a new record label, and you've been tasked with selecting the first three albums that will be added to the store, from a list of four.

The record label specializes in artists from the USA, and they have given Chinook some money to advertise the new albums in the USA, __so we're interested in finding out which genres sell the best in the USA.__

you should write up a summary of your findings, and make a recommendation for the three artists whose albums we should purchase for the store.

Submitted artists

| Artist Name | Genre | 
| --- | --- | --- |
| Regal | Hip-Hop | 
| Red Tone | Punk | 
| Meteor and the Girls  | Pop | 
| Slim Jim Bites  | Blues | 


In [7]:
q = "SELECT * FROM ALBUM"

run_query(q)

album_id                                              title  artist_id
0           1              For Those About To Rock We Salute You          1
1           2                                  Balls to the Wall          2
2           3                                  Restless and Wild          2
3           4                                  Let There Be Rock          1
4           5                                           Big Ones          3
5           6                                 Jagged Little Pill          4
6           7                                           Facelift          5
7           8                                     Warner 25 Anos          6
8           9                     Plays Metallica By Four Cellos          7
9          10                                         Audioslave          8
10         11                                       Out Of Exile          8
11         12                                BackBeat Soundtrack          9
12         13                           The Best Of Billy Cobham         10
13         14           Alcohol Fueled Brewtality Live! [Disc 1]         11
14         15           Alcohol Fueled Brewtality Live! [Disc 2]         11
15         16                                      Black Sabbath         12
16         17                    Black Sabbath Vol. 4 (Remaster)         12
17         18                                         Body Count         13
18         19                                   Chemical Wedding         14
19         20   The Best Of Buddy Guy - The Millenium Collection         15
20         21                                       Prenda Minha         16
21         22                              Sozinho Remix Ao Vivo         16
22         23                                     Minha Historia         17
23         24                                     Afrociberdelia         18
24         25                                    Da Lama Ao Caos         18
25         26                                Acústico MTV [live]         19
26         27                                Cidade Negra - Hits         19
27         28                                           Na Pista         20
28         29                                     Axé Bahia 2001         21
29         30                       BBC Sessions [Disc 1] [live]         22
..        ...                                                ...        ...
317       318                           SCRIABIN: Vers la flamme        250
318       319  Armada: Music from the Courts of England and S...        251
319       320                    Mozart: Symphonies Nos. 40 & 41        248
320       321                                      Back to Black        252
321       322                                              Frank        252
322       323              Carried to Dust (Bonus Track Version)        253
323       324          Beethoven: Symphony No. 6 'Pastoral' Etc.        254
324       325                   Bartok: Violin & Viola Concertos        255
325       326             Mendelssohn: A Midsummer Night's Dream        256
326       327                 Bach: Orchestral Suites Nos. 1 - 4        257
327       328      Charpentier: Divertissements, Airs & Concerts        258
328       329                             South American Getaway        259
329       330                            Górecki: Symphony No. 3        260
330       331                           Purcell: The Fairy Queen        261
331       332                      The Ultimate Relexation Album        262
332       333                  Purcell: Music for the Queen Mary        263
333       334                       Weill: The Seven Deadly Sins        264
334       335  J.S. Bach: Chaconne, Suite in E Minor, Partita...        265
335       336  Prokofiev: Symphony No.5 & Stravinksy: Le Sacr...        248
336       337                   Szymanowski: Piano Works, Vol. 1        266
337       338                        Nielsen: The Six

## Write a query that returns each genre, with the number of tracks sold in the USA:

In [8]:
q = '''SELECT 
    genre_id,
    COUNT(track_id) AS jumlah
FROM track T
GROUP BY genre_id
ORDER BY jumlah DESC'''

run_query(q)

genre_id  jumlah
0          1    1297
1          7     579
2          3     374
3          4     332
4          2     130
5         19      93
6          6      81
7         24      74
8         21      64
9         14      61
10         8      58
11         9      48
12        10      43
13        23      40
14        17      35
15        15      30
16        13      28
17        16      28
18        20      26
19        12      24
20        22      17
21        11      15
22        18      13
23         5      12
24        25       1

In [9]:
# SELECT SUM(QUANTITY)
# FROM INVOICE_line

q = "SELECT SUM(QUANTITY) FROM invoice_line"
run_query(q)

SUM(QUANTITY)
0           4757

In [10]:
q = "SELECT * FROM invoice WHERE billing_country = 'USA' "
run_query(q)

invoice_id  customer_id         invoice_date            billing_address  \
0             1           18  2017-01-03 00:00:00               627 Broadway   
1             4           18  2017-01-06 00:00:00               627 Broadway   
2             5           27  2017-01-07 00:00:00            1033 N Park Ave   
3             9           18  2017-01-18 00:00:00               627 Broadway   
4            14           25  2017-01-23 00:00:00      319 N. Frances Street   
5            17           20  2017-01-28 00:00:00       541 Del Medio Avenue   
6            18           24  2017-02-02 00:00:00      162 E Superior Street   
7            42           19  2017-03-18 00:00:00            1 Infinite Loop   
8            43           22  2017-03-19 00:00:00           120 S Orange Ave   
9            44           21  2017-03-20 00:00:00           801 W 4th Street   
10           50           18  2017-04-02 00:00:00               627 Broadway   
11           51           19  2017-04-03 00:00:00            1 Infinite Loop   
12           52           19  2017-04-11 00:00:00            1 Infinite Loop   
13           54           26  2017-04-13 00:00:00        2211 W Berry Street   
14           57           19  2017-04-22 00:00:00            1 Infinite Loop   
15           66           16  2017-04-29 00:00:00  1600 Amphitheatre Parkway   
16           74           26  2017-05-18 00:00:00        2211 W Berry Street   
17           75           16  2017-05-24 00:00:00  1600 Amphitheatre Parkway   
18           83           22  2017-06-15 00:00:00           120 S Orange Ave   
19           89           21  2017-06-23 00:00:00           801 W 4th Street   
20           98           17  2017-07-23 00:00:00            1 Microsoft Way   
21          101           20  2017-07-31 00:00:00       541 Del Medio Avenue   
22          109           22  2017-08-29 00:00:00           120 S Orange Ave   
23          118           28  2017-09-10 00:00:00                302 S 700 E   
24          128           17  2017-10-12 00:00:00            1 Microsoft Way   
25          129           17  2017-10-16 00:00:00            1 Microsoft Way   
26          139           23  2017-11-03 00:00:00            69 Salem Street   
27          143           25  2017-11-10 00:00:00      319 N. Frances Street   
28          145           28  2017-11-22 00:00:00                302 S 700 E   
29          147           19  2017-11-28 00:00:00            1 Infinite Loop   
..          ...          ...                  ...                        ...   
101         490           19  2020-03-14 00:00:00            1 Infinite Loop   
102         494           26  2020-03-31 00:00:00        2211 W Berry Street   
103         496           21  2020-04-01 00:00:00           801 W 4th Street   
104         499           17  2020-04-03 00:00:00            1 Microsoft Way   
105         505           19  2020-04-14 00:00:00            1 Infinite Loop   
106         510           25  2020-04-28 00:00:00      319 N. Frances Street   
107         511           24  2020-05-01 00:00:00      162 E Superior Street   
108         514           25  2020-05-14 00:00:00      319 N. Frances Street   
109         515           23  2020-05-15 00:00:00            69 Salem Street   
110         516           23  2020-05-16 00:00:00            69 Salem Street   
111         518           22  2020-05-22 00:00:00           120 S Orange Ave   
112         522           22  2020-05-27 00:00:00           120 S Orange Ave   
113         529           24  2020-06-12 00:00:00      162 E Superior Street   
114         543           17  2020-07-23 00:00:00            1 Microsoft Way   
115         551           16  2020-08-18 00:00:00  1600 Amphitheatre Parkway   
116         556           21  2020-09-04 00:00:00           801 W 4th Street   
117         558           17  2020-09-11 00:00:00            1 Microsoft Way   
118         561           23  2020-09-12 00:00:00            69 Salem Street   
119  

In [11]:
q = '''WITH invoice_usa AS(
    SELECT *
    FROM invoice
    WHERE billing_country = 'USA'
    )
    SELECT 
        G.genre_id,
        G.name,
        SUM(IL.quantity) as Jumlah_Album_Terjual,
        CAST(SUM(IL.quantity) AS FLOAT) / (   
            SELECT SUM(IL.quantity) 
            FROM invoice_line IL
            INNER JOIN invoice_usa iu2 ON IL.invoice_id = Iu2.invoice_id
             )*100 AS Percentage
    FROM invoice_usa IU
    INNER JOIN invoice_line IL ON IU.invoice_id = IL.invoice_id
    LEFT JOIN track T ON  IL.track_id = T.track_id 
    LEFT JOIN genre G ON T.genre_id = G.genre_id 
    GROUP BY G.genre_id
    ORDER BY Jumlah_Album_Terjual DESC'''

run_query(q)

genre_id                name  Jumlah_Album_Terjual  Percentage
0          1                Rock                   561   53.377735
1          4  Alternative & Punk                   130   12.369172
2          3               Metal                   124   11.798287
3         14            R&B/Soul                    53    5.042816
4          6               Blues                    36    3.425309
5         23         Alternative                    35    3.330162
6          7               Latin                    22    2.093245
7          9                 Pop                    22    2.093245
8         17         Hip Hop/Rap                    20    1.902950
9          2                Jazz                    14    1.332065
10        12      Easy Listening                    13    1.236917
11         8              Reggae                     6    0.570885
12        15   Electronica/Dance                     5    0.475737
13        24           Classical                     4    0.380590
14        13         Heavy Metal                     3    0.285442
15        10          Soundtrack                     2    0.190295
16        19            TV Shows                     1    0.095147


| Artist Name | Genre | 
| --- | --- | --- |
| Regal | Hip-Hop | 
| Red Tone | Punk | 
| Meteor and the Girls  | Pop | 
| Slim Jim Bites  | Blues | 


## A recommendation for the three artists whose albums we should purchase for the store, based on sales of tracks from their genres.

1. Red Tone - Punk
2. Slim Jim Bites - Blues
3. Meteor and te Girls - Pop

Rock Genre is the most genre that have been sold in our store. It reach more than half sales. Followed by 2 similar genres like Metal and Punk. 

So the most potential one is "Red Tone" with their Punk genre album.
Then, followed by "Slim Jim Bites" with Blues genre in fifth position of all, and "Meteor and te Girls" with Pop genre in eighth of all.

<hr>

# Case 2 : Analyzing Employee Sales Performance

Each customer for the Chinook store gets assigned to a sales support agent within the company when they first make a purchase.

__Analyze the purchases of customers belonging to each employee to see if any sales support agent is performing either better or worse than the others.__

In [12]:
q = "SELECT * FROM EMPLOYEE"

run_query(q)

employee_id last_name first_name                title  reports_to  \
0            1     Adams     Andrew      General Manager         NaN   
1            2   Edwards      Nancy        Sales Manager         1.0   
2            3   Peacock       Jane  Sales Support Agent         2.0   
3            4      Park   Margaret  Sales Support Agent         2.0   
4            5   Johnson      Steve  Sales Support Agent         2.0   
5            6  Mitchell    Michael           IT Manager         1.0   
6            7      King     Robert             IT Staff         6.0   
7            8  Callahan      Laura             IT Staff         6.0   

             birthdate            hire_date                      address  \
0  1962-02-18 00:00:00  2016-08-14 00:00:00          11120 Jasper Ave NW   
1  1958-12-08 00:00:00  2016-05-01 00:00:00                 825 8 Ave SW   
2  1973-08-29 00:00:00  2017-04-01 00:00:00                1111 6 Ave SW   
3  1947-09-19 00:00:00  2017-05-03 00:00:00             683 10 Street SW   
4  1965-03-03 00:00:00  2017-10-17 00:00:00                 7727B 41 Ave   
5  1973-07-01 00:00:00  2016-10-17 00:00:00         5827 Bowness Road NW   
6  1970-05-29 00:00:00  2017-01-02 00:00:00  590 Columbia Boulevard West   
7  1968-01-09 00:00:00  2017-03-04 00:00:00                  923 7 ST NW   

         city state country postal_code              phone                fax  \
0    Edmonton    AB  Canada     T5K 2N1  +1 (780) 428-9482  +1 (780) 428-3457   
1     Calgary    AB  Canada     T2P 2T3  +1 (403) 262-3443  +1 (403) 262-3322   
2     Calgary    AB  Canada     T2P 5M5  +1 (403) 262-3443  +1 (403) 262-6712   
3     Calgary    AB  Canada     T2P 5G3  +1 (403) 263-4423  +1 (403) 263-4289   
4     Calgary    AB  Canada     T3B 1Y7   1 (780) 836-9987   1 (780) 836-9543   
5     Calgary    AB  Canada     T3B 0C5  +1 (403) 246-9887  +1 (403) 246-9899   
6  Lethbridge    AB  Canada     T1K 5N8  +1 (403) 456-9986  +1 (403) 456-8485   
7  Lethbridge    AB  Canada     T1H 1Y8  +1 (403) 467-3351  +1 (403) 467-8772   

                      email  
0    andrew@chinookcorp.com  
1     nancy@chinookcorp.com  
2      jane@chinookcorp.com  
3  margaret@chinookcorp.com  
4     steve@chinookcorp.com  
5   michael@chinookcorp.com  
6    robert@chinookcorp.com  
7     laura@chinookcorp.com

In [13]:
q = '''SELECT 
    E.employee_id , 
    E.first_name AS EMPLOYEE, 
    C.customer_id, 
    C.first_name AS CUSTOMER, 
    SUM(I.total) Total
FROM employee E
LEFT JOIN customer C ON C.support_rep_id = E.employee_id
LEFT JOIN invoice I ON I.customer_id = C.customer_id
GROUP BY E.employee_id
ORDER BY Total DESC'''

# q = "SELECT E.employee_id , E.first_name AS EMPLOYEE, E.title, COUNT(C.customer_id) AS Jumlah_Customer, SUM(I.total) AS Total, AVG(I.total) as Average FROM employee E LEFT JOIN customer C ON C.support_rep_id = E.employee_id LEFT JOIN invoice I ON I.customer_id = C.customer_id GROUP BY E.employee_id ORDER BY Total DESC"

EMPLOYEE_RANK = run_query(q)
EMPLOYEE_RANK

employee_id  EMPLOYEE  customer_id CUSTOMER    Total
0            3      Jane         59.0     Puja  1731.51
1            4  Margaret         56.0    Diego  1584.00
2            5     Steve         57.0     Luis  1393.92
3            1    Andrew          NaN     None      NaN
4            2     Nancy          NaN     None      NaN
5            6   Michael          NaN     None      NaN
6            7    Robert          NaN     None      NaN
7            8     Laura          NaN     None      NaN

Dari hasil perolehan di atas, hanya ada 3 dari 8 pegawai yang berhubungan dengan customer, sesuai dengan jumlah pegawai dengan title "Sales Support Agent".

From the results above, there are only 3 out of 8 employees associated with customers, according to the number of employees with the title "Sales Support Agent".

<hr>

Dari 3 pegawai bagian sales tersebut, __Jane__ merupakan pegawai dengan total sales terbesar, dengan akuisisi customer nomor 2. Jane berhasil menghandle $1731.51 sales dengan jumlah customer sebesar 212.

From the 3 sales employees, __Jane__ is the employee with the largest total sales, with customer acquisition ranked 2. Jane managed to handle $ 1731.51 sales with a total customer of 212.

<hr>

Dan, __Margaret__ adalah pegawai dengan akuisisi customer terbanyak, dan sales terbanyak nomor 2. Margaret berhasil menghandle 214 customer dengan jumlah sales $1584.

And, __Margaret__ is the employee with the most customer acquisitions, and the most sales is number 2. Margaret managed to handle 214 customers with a total sales of $ 1584.

<hr>

And, the last position is Steve with only handle 188 customers and $1393.82 total sales. 

And, the last position is Steve with only handle 188 customers and $ 1393.82 total sales.

<hr>

In [44]:
q = '''
    SELECT
        first_name,
        title,
        hire_date
    FROM employee
    WHERE title = "Sales Support Agent" 
    ORDER BY hire_date

'''

run_query(q)

first_name                title            hire_date
0       Jane  Sales Support Agent  2017-04-01 00:00:00
1   Margaret  Sales Support Agent  2017-05-03 00:00:00
2      Steve  Sales Support Agent  2017-10-17 00:00:00

from the above table now we know that Jane and Margaret are hired 

<hr>

# Case 3 : Analyzing Sales Data by Country



In [14]:
q = '''
    SELECT country, count(customer_id) as cc
    FROM customer
    GROUP BY country
    order by cc desc
'''

run_query(q)

country  cc
0              USA  13
1           Canada   8
2           Brazil   5
3           France   5
4          Germany   4
5   United Kingdom   3
6   Czech Republic   2
7            India   2
8         Portugal   2
9        Argentina   1
10       Australia   1
11         Austria   1
12         Belgium   1
13           Chile   1
14         Denmark   1
15         Finland   1
16         Hungary   1
17         Ireland   1
18           Italy   1
19     Netherlands   1
20          Norway   1
21          Poland   1
22           Spain   1
23          Sweden   1

In [15]:
q = '''
    WITH customer_o AS(
    SELECT
        *,
        CASE
            WHEN customer.country IN (
                SELECT C.country
                FROM customer C 
                GROUP BY C.country
                HAVING COUNT(C.customer_id) = 1
            )
            THEN "Other"
            ELSE customer.country
            END AS Country_edited
    FROM customer)
    SELECT 
        count(*) COUNT_TRANSAKSI,
        count(I.invoice_id) COUNT_invoice
    FROM customer_o CO
    INNER JOIN invoice I ON CO.customer_id = I.customer_id
    GROUP BY CO.country_edited
    
    '''

run_query(q)

COUNT_TRANSAKSI  COUNT_invoice
0               61             61
1               76             76
2               30             30
3               50             50
4               41             41
5               21             21
6              147            147
7               29             29
8              131            131
9               28             28

Sales by country analysis

Country which has only one customer, grouped in "other". Then, we show each country data about : 
- "total number of cust" 
- "total of sales"
- "avg value of sales per customer"
- "average order value"

In [16]:
q = '''
    WITH customer_o AS(
        SELECT
            *,
            CASE
                WHEN customer.country IN (
                    SELECT C.country
                    FROM customer C 
                    GROUP BY C.country
                    HAVING COUNT(C.customer_id) = 1
                )
                THEN "Other"
                ELSE customer.country
                END AS Country_edited
        FROM customer)
        
    
    
    SELECT 
        CO.Country_edited AS Country,
        count(*) AS Total_transaction,
        CC.count_customer AS total_customer,
        SUM(I.total) AS Total_sales,
        SUM(I.total)/CC.count_customer AS Avg_sales_percustomer,
        AVG(I.total) AS Avg_order_value,
        CASE 
            WHEN CO.country_edited = 'Other' 
                THEN 1
            ELSE 0
            END AS Sort
    FROM customer_o CO
    INNER JOIN invoice I ON CO.customer_id = I.customer_id
    INNER JOIN (
        SELECT
            country_edited,
            COUNT(customer_id) as count_customer
        FROM customer_o
        GROUP BY country_edited
    ) AS CC ON CO.country_edited = CC.country_edited
    GROUP BY CO.country_edited
    ORDER BY Sort, Total_sales DESC
    '''

run_query(q)

Country  Total_transaction  total_customer  Total_sales  \
0             USA                131              13      1040.49   
1          Canada                 76               8       535.59   
2          Brazil                 61               5       427.68   
3          France                 50               5       389.07   
4         Germany                 41               4       334.62   
5  Czech Republic                 30               2       273.24   
6  United Kingdom                 28               3       245.52   
7        Portugal                 29               2       185.13   
8           India                 21               2       183.15   
9           Other                147              15      1094.94   

   Avg_sales_percustomer  Avg_order_value  Sort  
0              80.037692         7.942672     0  
1              66.948750         7.047237     0  
2              85.536000         7.011148     0  
3              77.814000         7.781400     0  
4              83.655000         8.161463     0  
5             136.620000         9.108000     0  
6              81.840000         8.768571     0  
7              92.565000         6.383793     0  
8              91.575000         8.721429     0  
9              72.996000         7.448571     1

USA memiliki total sales tertinggi, dan total customer tertinggi. Tim marketing dapat memaksimalkan iklan dan promo kepada customer yang sudah ada di USA untuk lebih meningkatkan sales percustomer serta jumlah penjualan pertransaksinya.



<hr>

# Case 4 : Albums vs Individual Tracks

bfind out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue.

query that categorizes each invoice as either an album purchase or not

In [17]:
q = '''
WITH invoice_album_p AS (
    SELECT IL.invoice_id, T.album_id,
    CASE
        WHEN sum(T.album_id)/count(T.album_id) != T.album_id
        THEN 0
        WHEN sum(T.album_id)/count(T.album_id) = T.album_id 
            AND count(T.album_id) <= 2
        THEN -1
        WHEN sum(T.album_id)/count(T.album_id) = T.album_id 
        THEN 1
        ELSE 0 
        END AS Album_purchase
    FROM invoice_line IL
    INNER JOIN track T ON T.track_id = IL.track_id
    GROUP BY IL.invoice_id
    )
    SELECT 
        COUNT(I.invoice_id) as Number_of_invoices,
        CAST(COUNT(I.invoice_id) AS FLOAT)/(
            SELECT count(I2.invoice_id)
            FROM invoice I2
        )* 100 AS Percentage_of_invoice,
        IAP.Album_purchase
    FROM invoice I
    INNER JOIN invoice_album_p IAP ON IAP.invoice_id = I.invoice_id
    GROUP BY IAP.Album_purchase
'''

run_query(q)

Number_of_invoices  Percentage_of_invoice  Album_purchase
0                  38               6.188925              -1
1                 442              71.986971               0
2                 134              21.824104               1

Based on previous sales, there are only about 21.8% album-purchased-type invoices.
So, the most invoices are track-type purchases. 

The Chinook Store may continue to buy full albums, because we think about 20% revenue that will be lost if the stor stop it.

<hr>

<hr>

other exploration

## Which artist is used in the most playlists?

In [18]:
q = '''
WITH playlist_and_artist AS (
    SELECT
        P.playlist_id,
        PT.playlist_id,
        P.name as playlist_name,
        T.track_id,
        T.album_id,
        AR.artist_id,
        AR.name as artist_name
    FROM playlist P
    INNER JOIN playlist_track PT ON PT.playlist_id = P.playlist_id
    LEFT JOIN track T ON T.track_id = PT.track_id
    LEFT JOIN album A ON A.album_id = T.album_id
    LEFT JOIN artist AR ON AR.artist_id = A.artist_id

    GROUP BY A.artist_id, P.playlist_id
    ORDER BY A.artist_id
    )
    SELECT
        PAA.artist_id,
        PAA.artist_name,
        COUNT(PAA.playlist_id) as number_of_playlist
        
    FROM playlist_and_artist PAA
    GROUP BY PAA.artist_id
    ORDER BY number_of_playlist DESC
    LIMIT 6
'''
     
run_query(q)

artist_id                                        artist_name  \
0        226                                     Eugene Ormandy   
1        208                   English Concert & Trevor Pinnock   
2        214  Academy of St. Martin in the Fields & Sir Nevi...   
3        247                                 The King's Singers   
4        248      Berliner Philharmoniker & Herbert Von Karajan   
5        206             Alberto Turco & Nova Schola Gregoriana   

   number_of_playlist  
0                   7  
1                   6  
2                   6  
3                   6  
4                   6  
5                   5

Above is the top artist who used in the most playlist.
The first rank is Eugene Ormandy, appear in 7 number of playlists, folowwed by 4 other artist that appear in 6 playlists.

<hr>

## How many tracks have been purchased vs not purchased?

In [19]:
q = '''
SELECT count(track_id)
FROM track

'''

run_query(q)

count(track_id)
0             3503

total number of track is about 3503 tracks

In [20]:
q = '''
    SELECT
        T.track_id,
        CASE 
            WHEN IL.invoice_line_id IS NULL
            THEN 0
            ELSE 1
            END AS Purchased_track
    FROM track T
    LEFT JOIN invoice_line IL ON IL.track_id = T.track_id
    GROUP BY  T.track_id
    ORDER BY IL.invoice_line_id 
'''

run_query(q)

track_id  Purchased_track
0           99                0
1          101                0
2          104                0
3          106                0
4          107                0
5          111                0
6          112                0
7          113                0
8          114                0
9          115                0
10         116                0
11         117                0
12         118                0
13         119                0
14         120                0
15         121                0
16         122                0
17         123                0
18         124                0
19         125                0
20         127                0
21         129                0
22         133                0
23         136                0
24         141                0
25         143                0
26         144                0
27         145                0
28         148                0
29         159                0
...        ...              ...
3473      2208                1
3474      2209                1
3475      2210                1
3476      2211                1
3477      2212                1
3478      2213                1
3479      2214                1
3480      2215                1
3481      1865                1
3482      2647                1
3483      3292                1
3484      2731                1
3485      3465                1
3486      2948                1
3487      2738                1
3488      1126                1
3489      1915                1
3490      2650                1
3491      2651                1
3492      2652                1
3493      2654                1
3494      2655                1
3495      2656                1
3496      2657                1
3497      2658                1
3498      2659                1
3499      2660                1
3500      2661                1
3501      2662                1
3502      2663                1

[3503 rows x 2 columns]

Above is the table shows the track and its status whether purchased or not. We will use this kind of table to subqueries below

In [21]:
q = '''
WITH track_purchased_table AS(
    SELECT
        T.track_id,
        CASE 
            WHEN IL.invoice_line_id IS NULL
            THEN 0
            ELSE 1
            END AS Purchased_track
    FROM track T
    LEFT JOIN invoice_line IL ON IL.track_id = T.track_id
    GROUP BY  T.track_id
    ORDER BY IL.invoice_line_id 
)
SELECT 
    TPT.Purchased_track,
    count(TPT.track_id) AS number_of_track,
    count(TPT.track_id)/ 
    CAST((SELECT COUNT(track_id) FROM track) AS FLOAT)*100 AS Percentage
    
FROM track_purchased_table TPT
GROUP BY TPT.Purchased_track
'''

run_query(q)

Purchased_track  number_of_track  Percentage
0                0             1697   48.444191
1                1             1806   51.555809

There are 1697 tracks have not been purchased yet, and this is a big number. It about 48 percent of all. 

We recommend the chinook store to evaluate how they provide the tracks. Chinook may concern to the kind of tracks that customer need and love.

And it may be about how they promote the unpurchased tracks too. The chinook store may recommend the tracks which similar with the customer's purchased track or based on another consideration.

## Do protected vs non-protected media types have an effect on popularity?

In [22]:
q ='''
    SELECT *
    FROM media_type
    GROUP by name

'''

run_query(q)

media_type_id                         name
0              5               AAC audio file
1              1              MPEG audio file
2              2     Protected AAC audio file
3              3  Protected MPEG-4 video file
4              4     Purchased AAC audio file

Above is the table of media type, there are 2 type of protected type. 

In [26]:
q ='''
    SELECT media_type_id
    FROM media_type
    WHERE LOWER(name) LIKE '%protected%'
    GROUP by name
    
    '''
run_query(q)

media_type_id
0              2
1              3

2 and 3 are the ID of media type that indicates "protected" type.

In [29]:
q = '''
    SELECT 
    T.track_id,
    SUM(IL.quantity) AS Qty,
    CASE  
        WHEN T.media_type_id IN (
            SELECT media_type_id
            FROM media_type
            WHERE LOWER(name) LIKE '%protected%'
            GROUP by name
        )
        THEN 1
        ELSE 0
        END AS protected
    
    FROM track T
    LEFT JOIN invoice_line IL ON IL.track_id = T.track_id
    GROUP BY T.track_id
    ORDER BY Qty DESC

'''

run_query(q)

track_id   Qty  protected
0         3336  31.0          0
1         1489  14.0          0
2         1495  14.0          0
3            6  13.0          0
4         1487  13.0          0
5         1490  13.0          0
6         1483  12.0          0
7         2558  12.0          0
8         1129  11.0          0
9         1479  11.0          0
10        1485  11.0          0
11        1488  11.0          0
12        1491  11.0          0
13        1493  11.0          0
14        2563  11.0          0
15          13  10.0          0
16        1122  10.0          0
17        1126  10.0          0
18        1466  10.0          0
19        1482  10.0          0
20        1484  10.0          0
21        1492  10.0          0
22        2234  10.0          0
23        2235  10.0          0
24        2561  10.0          0
25        2640  10.0          0
26        2663  10.0          0
27        3478  10.0          1
28           9   9.0          0
29          46   9.0          0
...        ...   ...        ...
3473      3443   NaN          1
3474      3444   NaN          1
3475      3445   NaN          1
3476      3446   NaN          1
3477      3447   NaN          1
3478      3449   NaN          1
3479      3450   NaN          1
3480      3451   NaN          1
3481      3453   NaN          1
3482      3454   NaN          1
3483      3479   NaN          0
3484      3481   NaN          1
3485      3482   NaN          1
3486      3483   NaN          1
3487      3485   NaN          1
3488      3486   NaN          1
3489      3487   NaN          1
3490      3488   NaN          1
3491      3489   NaN          1
3492      3490   NaN          1
3493      3492   NaN          1
3494      3493   NaN          1
3495      3494   NaN          1
3496      3496   NaN          0
3497      3497   NaN          1
3498      3498   NaN          0
3499      3499   NaN          1
3500      3501   NaN          1
3501      3502   NaN          1
3502      3503   NaN          1

[3503 rows x 3 columns]

Above, we try to know which is the most track have been purchased, and it shows that the most top qty have not protected media type

In [35]:
q ='''
WITH Track_qty_protected AS(
    SELECT 
    T.track_id,
    SUM(IL.quantity) AS Qty,
    CASE  
        WHEN T.media_type_id IN (
            SELECT media_type_id
            FROM media_type
            WHERE LOWER(name) LIKE '%protected%'
            GROUP by name
        )
        THEN 1
        ELSE 0
        END AS protected
    
    FROM track T
    LEFT JOIN invoice_line IL ON IL.track_id = T.track_id
    GROUP BY T.track_id
    ORDER BY Qty DESC
)
SELECT 
    protected,
    SUM(Qty) as Total_Qty,
    CAST(SUM(Qty) AS FLOAT)/(
        SELECT SUM(Qty) FROM Track_qty_protected
    ) * 100
    as percentage
FROM Track_qty_protected
GROUP BY protected

'''

run_query(q)

protected  Total_Qty  percentage
0          0       4315    90.70843
1          1        442     9.29157

from the table above, we know that 90% of purchased tracks are not protected.
